pyteal/ast/scratchvar.py

```diff
@@ -47,5 +54,79 @@ def load(self) -> ScratchLoad:
         """Load value from Scratch Space"""
         return self.slot.load(self.type)
 
+    def index(self) -> Expr:
+        return self.slot.index()
+
 
 ScratchVar.__module__ = "pyteal"
+
+
+class DynamicScratchVar(ScratchVar):
```

Comment:
`DynamicCratchVar` is now a `ScratchVar`

Reply:
Link to the prior discussion: https://github.com/algorand/pyteal/pull/198#discussion_r810271366

Reply:
@jasonpaulos Thanks for the pointer.  

With the disclaimer that I still have gaps in my intuition, here's my thinking:
* Is it accurate to say that `DynamicScratchVar` is _only_ intended for internal-to-the-compiler usage rather than also being used by PyTeal developers?
  * If also intended for PyTeal developers, I'd be interested in understanding the use cases. 
  * If intended for internal usage only, I'm wondering:  How can we more clearly delineate public vs private APIs?   
* With the assumption that `DynamicScratchVar` usage is _internal_, I feel like I'm _not_ convinced `DynamicScratchVar` _needs_ to extend `ScratchVar`.
  * Take my skepticism with a grain of salt.  I'm probably skeptical because there's gaps in my intuition.  I'd at least like to keep considering the topic.
  * As written, I see `evaluateSubroutine` in  `subroutine.py` depends on methods with the same signatures to invoke the subroutine.
    * I guess duck typing is at play here.  And it explains why _no_ tests fail when I removed the inheritance locally.
  * The skepticism arises for reasons similar to what @tzaffi raises. And my prior experiences bias me towards composition over inheritance.

Reply:
Summarizing verbal discussion below.

@tzaffi I'm good to resolve the thread on my side.  However, since I didn't open the thread, I leave it open in case you wish to add something.  Feel welcomed to resolve if you're good here too.

> Is it accurate to say that DynamicScratchVar is only intended for internal-to-the-compiler usage rather than also being used by PyTeal developers?

* `DynamicScratchVar` is intended for internal _and_ PyTeal developer usage.
* An example use case involves allowing a PyTeal developer to reserve an array of slots.  The cursor provided by `DynamicScratchVar` is a step towards providing such a facility.
* Given these points, I don't have further reason to suggest avoiding inheritance here.